### ISSUE TITLE
Hoist optimizer class out of base layer class into learning class

### ISSUE THREAD

**Issue by bvanessen** [LLNL/lbann#86]
I believe that we can hoist the optimizer out, since it should only be applied to the weight matrix.

**Comment by ndryden** [comment]
This is true in most cases, but for example, batchnorm has two optimizer instances to learn its parameters.

**Comment by bvanessen** [comment]
Okay, good point.

Brian C. Van Essen
vanessen1@llnl.gov
(w) 925-422-9300
(c) 925-290-5470

**Comment by bvanessen** [comment]
Doesn’t batch norm use its own self contained optimizers?  If they are hard coded as part of the actual layer, do they use the same approach to adding them as everyone else?

Brian C. Van Essen
vanessen1@llnl.gov
(w) 925-422-9300
(c) 925-290-5470

**Comment by ndryden** [comment]
Think this was clarified on the phone, but for everyone else: its optimizers are currently self-contained and currently constructed from the model's optimizer factory, but this is easy to change.

**Comment by bvanessen** [comment]
The optimizers have been hoisted into the learning class.  Batch normalization is the main exception and it already gets its optimizers via the model class.
